File: Cours/AP1/CH4/TD21/calculs.c
#include <math.h>
#include <stdio.h>

unsigned fact(unsigned n) {
    unsigned res;
    for (res = 1; n > 1; n--) {
        res *= n;
    }
    return res;
}

void test_fact(void) {
    for (unsigned i = 0; i < 6; i++) {
        printf("%u! = %u\n", i, fact(i));
    }
    puts("");
}

double racine(double x) {
    if (x < 0.0) {
        return -NAN;
    }
    double res = 1.0;
    for (unsigned nbr_iter = 10; nbr_iter > 0; nbr_iter--) {
        res = (res + (x / res)) / 2.0;
    }
    return res;
}

void test_racine(void) {
    for (double val = -1.0; val < 10.0; val += 1.0) {
        printf("racine(%g) = %g ~ %g\n", val, racine(val), sqrt(val));
    }
}

int pgcd(unsigned a, unsigned b){
    if ((a == 0) || (b==0)){
        return -1;
    }

    for (; a!=b;) {

        if (a > b) {
            a = a - b;
        }else{
            b = b - a;
        }       
    }

    return a;
}

void test_pgcd(void){
    printf("Test pgcd %d",pgcd(0, 0));
    printf("Test pgcd %d",pgcd(1, 0));
    printf("Test pgcd %d",pgcd(15, 10));
    printf("Test pgcd %d",pgcd(10, 15));
}

int main(void) {
    printf("Test");
    test_fact();
    test_racine();
    test_pgcd();
}

File: Cours/AP1/CH2/TD5/calculs/calculs.go
package main

import (
	"fmt"
	"math"
)

func somme(n int) int {

	if n < 0 {
		return 0
	} else {
		somme := 0
		for i := 0; i <= n; i++ {
			somme += i
		}

		return somme
	}

}

func testSomme() {
	for test := -1; test <= 5; test++ {
		fmt.Println(somme(test))
	}
}

func fact(n uint) uint {
	temp := 1
	if n == 0 || n == 1 {
		return 1
	}

	for i := n; i > 0; i-- {
		temp = temp * int(i)
	}

	return uint(temp)
}

func testFact() {
	for test := 0; test <= 5; test++ {
		fmt.Println(fact(uint(test)))
	}
}

func pgcd(a, b uint) int {
	if a == 0 || b == 0 {
		return -1
	}

	for {
		if a == b {
			break
		}

		if a > b {
			a = a - b
		} else {
			b = b - a
		}
	}

	return int(a)
}

func testPGCD() {
	fmt.Println(pgcd(120, 60))
	fmt.Println(pgcd(0, 12))
	fmt.Println(pgcd(50, 43))
}

func racine(x float64) float64 {
	if x < 0 {
		return math.NaN()
	}
	r_i := 1.0
	for i := 1; i <= 10; i++ {
		r_i = (r_i + x/r_i) / 2
	}

	return r_i

}

func testRacine() {
	for test := -1; test <= 9; test++ {
		fmt.Println(math.Sqrt(float64(test)), racine(float64(test)))
	}
}

func main() {
	testRacine()
}
